### PR TITLE
Generate the roles and role description

### PR DESCRIPTION
**Pull request by Appointat**
## Description

**Note**: Due to GitHub's security constraints, secret keys are inaccessible in CI/CD workflows for PRs raised from the [fork](https://github.com/Appointat/camel/tree/Role-Generation). This is to safeguard sensitive information. To address this and ensure the workflows run correctly, I created a new [branch](https://github.com/camel-ai/camel/tree/feature/role-generation), "feature/role-generation", directly under the original project and raised this PR from there. The previous [PR #240](https://github.com/camel-ai/camel/pull/240) from the fork will be **closed** to avoid confusion.

Describe your changes in detail.

## Motivation and Context

- [x] I have raised an issue to propose this change ([Feature Request](https://github.com/camel-ai/camel/issues/239 for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [x] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [x] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [x] I have updated the documentation accordingly.

## Output of the new example
``` shell
AI Assistant sys message:
BaseMessage(role_name='Software Engineer', role_type=<RoleType.ASSISTANT: 'assistant'>, meta_dict={'task': 'Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.', 'assistant_role': 'Software Engineer', 'user_role': 'Financial Analyst', 'assistant_description': "Designing, developing, and testing software for the trading bot; creating algorithms and data structures to optimize the bot's performance; troubleshooting and debugging the bot; monitoring the bot's performance and making adjustments as needed.", 'user_description': "Analyzing financial data and market trends; researching and forecasting stock prices; developing strategies for the trading bot; monitoring the bot's performance and making adjustments as needed."}, content="===== ROLES WITH DESCRIPTION =====\nFinancial Analyst and Software Engineer are collaborating to complete a task: Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.\nFinancial Analyst's competencies, professional characteristics, duties and workflows to complete the task: Analyzing financial data and market trends; researching and forecasting stock prices; developing strategies for the trading bot; monitoring the bot's performance and making adjustments as needed.\nSoftware Engineer's competencies, professional characteristics, duties and workflows to complete the task: Designing, developing, and testing software for the trading bot; creating algorithms and data structures to optimize the bot's performance; troubleshooting and debugging the bot; monitoring the bot's performance and making adjustments as needed.\n===== RULES OF ASSISTANT =====\nNever forget you are a Software Engineer and I am a Financial Analyst. Never flip roles! Never instruct me!\nWe share a common interest in collaborating to successfully complete a task.\nYou must help me to complete the task.\nHere is the task: Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.. Never forget our task!\nI must instruct you based on your expertise and my needs to complete the task.\n\nI must give you one instruction at a time.\nYou must write a specific solution that appropriately solves the requested instruction and explain your solutions.\nYou must decline my instruction honestly if you cannot perform the instruction due to physical, moral, legal reasons or your capability and explain the reasons.\nUnless I say the task is completed, you should always start with:\n\nSolution: <YOUR_SOLUTION>\n\n<YOUR_SOLUTION> should be very specific, include detailed explanations and provide preferable detailed implementations and examples and lists for task-solving.\nAlways end <YOUR_SOLUTION> with: Next request.")

AI User sys message:
BaseMessage(role_name='Financial Analyst', role_type=<RoleType.USER: 'user'>, meta_dict={'task': 'Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.', 'assistant_role': 'Software Engineer', 'user_role': 'Financial Analyst', 'assistant_description': "Designing, developing, and testing software for the trading bot; creating algorithms and data structures to optimize the bot's performance; troubleshooting and debugging the bot; monitoring the bot's performance and making adjustments as needed.", 'user_description': "Analyzing financial data and market trends; researching and forecasting stock prices; developing strategies for the trading bot; monitoring the bot's performance and making adjustments as needed."}, content='===== ROLES WITH DESCRIPTION =====\nFinancial Analyst and Software Engineer are collaborating to complete a task: Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.\nFinancial Analyst\'s competencies, professional characteristics, duties and workflows to complete the task: Analyzing financial data and market trends; researching and forecasting stock prices; developing strategies for the trading bot; monitoring the bot\'s performance and making adjustments as needed.\nSoftware Engineer\'s competencies, professional characteristics, duties and workflows to complete the task: Designing, developing, and testing software for the trading bot; creating algorithms and data structures to optimize the bot\'s performance; troubleshooting and debugging the bot; monitoring the bot\'s performance and making adjustments as needed.\n===== RULES OF USER =====\nNever forget you are a Financial Analyst and I am a Software Engineer. Never flip roles! You will always instruct me.\nWe share a common interest in collaborating to successfully complete a task.\nI must help you to complete the task.\nHere is the task: Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.. Never forget our task!\nYou must instruct me based on my expertise and your needs to solve the task ONLY in the following two ways:\n\n1. Instruct with a necessary input:\nInstruction: <YOUR_INSTRUCTION>\nInput: <YOUR_INPUT>\n\n2. Instruct without any input:\nInstruction: <YOUR_INSTRUCTION>\nInput: None\n\nThe "Instruction" describes a task or question. The paired "Input" provides further context or information for the requested "Instruction".\n\nYou must give me one instruction at a time.\nI must write a response that appropriately solves the requested instruction.\nI must decline your instruction honestly if I cannot perform the instruction due to physical, moral, legal reasons or my capability and explain the reasons.\nYou should instruct me not ask me questions.\nNow you must start to instruct me using the two ways described above.\nDo not add anything else other than your instruction and the optional corresponding input!\nKeep giving me instructions and necessary inputs until you think the task is completed.\nWhen the task is completed, you must only reply with a single word <CAMEL_TASK_DONE>.\nNever say <CAMEL_TASK_DONE> unless my responses have solved your task.')

Original task prompt:
Develop a trading bot for the stock market.

Specified task prompt:
Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.

Final task prompt:
Develop a highly efficient and self-learning trading bot that incorporates sentiment analysis algorithms to analyze social media trends in real-time, enabling Financial Analyst to make data-driven decisions in volatile market conditions, and integrate it seamlessly into the existing trading platform.

AI User: Financial Analyst

Instruction: Develop a sentiment analysis algorithm to analyze social media trends in real-time.
Input: None


AI Assistant:Software Engineer

Solution: To develop a sentiment analysis algorithm to analyze social media trends in real-time, we can follow the following steps:

1. Data Collection: Gather a large dataset of social media posts or tweets related to the financial market. This dataset should include both positive and negative sentiment examples.

2. Preprocessing: Clean the collected data by removing any irrelevant information such as URLs, hashtags, and special characters. Also, perform tokenization to split the text into individual words or tokens.

3. Feature Extraction: Convert the preprocessed text into numerical features that can be used by machine learning algorithms. Common techniques for feature extraction include bag-of-words, TF-IDF, and word embeddings like Word2Vec or GloVe.

4. Labeling: Manually label the collected data with sentiment labels (positive or negative). This step is necessary to train a supervised machine learning model.
(not finished...)
```

### CODE REVIEW

File: camel/societies/role_playing.py [hunk 1]
```diff
@@ -243,10 +269,20 @@ def get_sys_message_info(
         sys_msg_meta_dicts = [dict(task=self.task_prompt) for _ in range(2)]
         if (extend_sys_msg_meta_dicts is None and self.task_type
                 in [TaskType.AI_SOCIETY, TaskType.MISALIGNMENT]):
-            extend_sys_msg_meta_dicts = [
-                dict(assistant_role=assistant_role_name,
-                     user_role=user_role_name) for _ in range(2)
-            ]
+            if (assistant_description is not None
```
Comment (Obs01ete) [inline camel/societies/role_playing.py:272]: What if assistant_description is not None and user_description is None?

File: examples/test/test_ai_society_example.py [hunk 2]
```diff
@@ -23,6 +24,12 @@ def test_ai_society_role_playing_example():
         examples.ai_society.role_playing.main(ModelType.STUB)
 
 
+def test_role_playing_with_role_generation_example():
+    with patch('time.sleep', return_value=None):
+        examples.ai_society.role_playing_with_role_generation.main(
+            ModelType.GPT_3_5_TURBO)
```
Comment (Obs01ete) [inline examples/test/test_ai_society_example.py:30]: Make it with STUB only.

File: camel/generators.py [hunk 3]
```diff
@@ -29,13 +29,16 @@ class SystemMessageGenerator:
         sys_msg_meta_dict_keys (Optional[Set[str]], optional): The set of keys
             of the meta dictionary used to fill the prompts.
             (default: :obj:`None`)
+        with_role_description (bool, optional): Whether to include the role
```
Comment (Obs01ete) [inline camel/generators.py:32]: Test SystemMessageGenerator in both modes: without and with the role description.

File: camel/prompts/prompt_templates.py [hunk 4]
```diff
@@ -50,7 +57,14 @@ def get_prompt_from_key(self, task_type: TaskType, key: Any) -> TextPrompt:
                 :obj:`task_type` and :obj:`key`.
         """
         try:
-            return self.task_prompt_template_dict[task_type][key]
+            if (with_role_description):
```
Comment (Obs01ete) [inline camel/prompts/prompt_templates.py:60]: Remove brackets

File: camel/societies/role_playing.py [hunk 5]
```diff
@@ -121,12 +121,32 @@ def __init__(
...
+        assistant_description = (None if assistant_agent_kwargs is None else
+                                 assistant_agent_kwargs.get(
+                                     "role_description", None))
+        user_description = (None if user_agent_kwargs is None else
+                            user_agent_kwargs.get("role_description", None))
         (init_assistant_sys_msg, init_user_sys_msg,
          sys_msg_meta_dicts) = self.get_sys_message_info(
-             assistant_role_name, user_role_name, sys_msg_generator,
-             extend_sys_msg_meta_dicts)
+             assistant_role_name=assistant_role_name,
```
Comment (Obs01ete) [inline camel/societies/role_playing.py:144]: Remove `key=` from positional arguments.

**Review by Obs01ete — CHANGES_REQUESTED** [review]
Thanks for the feature. Let's polish it and have good code coverage.

File: camel/societies/role_playing.py [hunk 6]
```diff
@@ -243,10 +266,24 @@ def get_sys_message_info(
         sys_msg_meta_dicts = [dict(task=self.task_prompt) for _ in range(2)]
         if (extend_sys_msg_meta_dicts is None and self.task_type
                 in [TaskType.AI_SOCIETY, TaskType.MISALIGNMENT]):
-            extend_sys_msg_meta_dicts = [
-                dict(assistant_role=assistant_role_name,
-                     user_role=user_role_name) for _ in range(2)
-            ]
+            if (assistant_description is not None
```
Comment (Obs01ete) [inline camel/societies/role_playing.py:269]: This logic is getting out of control. Can we make a bird eye view on this and understand do we really mean that much complexity that results in such a bulky code?

File: examples/test/test_ai_society_example.py [hunk 7]
```diff
@@ -23,6 +25,17 @@ def test_ai_society_role_playing_example():
         examples.ai_society.role_playing.main(ModelType.STUB)
 
 
+def test_ai_society_role_assignment_example():
+    with patch('time.sleep', return_value=None):
+        examples.ai_society.role_assignment.main(ModelType.GPT_3_5_TURBO)
+
+
+def test_role_playing_with_role_generation_example():
```
Comment (Obs01ete) [inline examples/test/test_ai_society_example.py:33]: Why again having this white crow non-stub test?

**Review by Obs01ete — CHANGES_REQUESTED** [review]

File: camel/agents/role_assignment.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+class RoleAssignmentAgent(ChatAgent):
+    r"""
+    An agent that generates role names based on the task prompt.
+    Attributes:
+        role_assignment_prompt (TextPrompt): A prompt for the agent to generate
+        role names.
+    args:
+        model (ModelType): The tupe of model to use for the agent.
+            (default: :obj: 'ModelType.GPT_3_5_TURBO')
```
Comment (lightaime) [inline camel/agents/role_assignment.py:33]:
```suggestion
        model (ModelType, optional): The type of model to use for the agent.
            (default: :obj:`ModelType.GPT_3_5_TURBO`)
```

File: camel/agents/role_assignment.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        model_config (Any): The configuration for the model.
+            (default: :obj: 'None')
```
Comment (lightaime) [inline camel/agents/role_assignment.py:35]:
```suggestion
        model_config (Any, optional): The configuration for the model.
            (default: :obj:`None`)
```

File: camel/agents/role_assignment.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    """
+
+    def __init__(
+        self,
+        model: ModelType = ModelType.GPT_3_5_TURBO,
+        model_config: Optional[Any] = None,
+    ) -> None:
+        self.role_assignment_prompt = TextPrompt(
+            'Given this task, "{task}", generate two role names, ' +
```
Comment (lightaime) [inline camel/agents/role_assignment.py:44]: ?

File: camel/agents/role_assignment.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+            role_type=RoleType.ASSISTANT,
+            meta_dict=None,
+            content="You assign roles based on tasks.",
+        )
+        super().__init__(system_message, model, model_config)
+
+    @retry(wait=wait_exponential(min=5, max=60), stop=stop_after_attempt(5))
+    def run_role_with_description(
+        self,
+        num_roles: Optional[int] = 2,
```
Comment (lightaime) [inline camel/agents/role_assignment.py:58]:
```suggestion
        num_roles: int = 2,
```

File: camel/agents/role_assignment.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+        task_prompt: Union[str, TextPrompt] = "",
+    ) -> Tuple[List[str], Dict[str, str], bool, Dict[str, Any]]:
+        r""" "
+        Generate role names based on the input task prompt.
```
Comment (lightaime) [inline camel/agents/role_assignment.py:62]:
```suggestion
        r"""Generate role names based on the input task prompt.
```

File: camel/agents/role_assignment.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+        Args:
+            num_roles (int): The number of roles to generate.
```
Comment (lightaime) [inline camel/agents/role_assignment.py:65]:
```suggestion
            num_roles (int, optional): The number of roles to generate.
```

File: camel/agents/role_assignment.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+                (default: :obj:`2`)
+            task_prompt (Union[str, TextPrompt]): The prompt
+                for the task based on which the roles are to be generated.
```
Comment (lightaime) [inline camel/agents/role_assignment.py:68]: I guess the `task_prompt` is required. Should we put it as a positional argument?

File: camel/prompts/ai_society.py
Comment (lightaime) [inline camel/prompts/ai_society.py:1]: Prefer to create a new file for this PR and compose it with the prompts from `ai_society.py`. An example can be found in `misalignment.py`. Prefer to remove all the `with_role_description ` flags and create a new `ASSISTANT_PROMPT` and `USER_PROMPT` with `ROLE_DESCRIPTION_PROMPT` in it.

**Review by lightaime — COMMENTED** [review]
Thanks for the PR! I think all the `with_role_description` flags should be removed and just create new `ASSISTANT_PROMPT` and `USER_PROMPT` prompts with role_description in it.

File: examples/role_description/role_playing_with_role_description.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    ]
+
+    role_play_session = RolePlaying(
+        task_prompt=task_prompt,
+        task_type=TaskType.ROLE_DESCRIPTION,  # important for role description
+        with_task_specify=True,
+        assistant_role_name=ai_assistant_role,
+        user_role_name=ai_user_role,
+        task_specify_agent_kwargs=dict(model=model_type),
+        extend_sys_msg_meta_dicts=sys_msg_meta_dicts,
```
Comment (lightaime) [inline examples/role_description/role_playing_with_role_description.py:54]:
```suggestion
        assistant_role_name=ai_assistant_role,
        user_role_name=ai_user_role,
        task_prompt=task_prompt,
        task_type=TaskType.ROLE_DESCRIPTION,  # important for role description
        with_task_specify=True,
        task_specify_agent_kwargs=dict(model=model_type),
        extend_sys_msg_meta_dicts=sys_msg_meta_dicts,
```

File: examples/test/test_role_description.py [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+def test_role_generation_example():
+    with patch('time.sleep', return_value=None):
+        examples.role_description.role_generation.main(ModelType.GPT_3_5_TURBO)
+
+
+def test_role_playing_with_role_description_example():
+    with patch('time.sleep', return_value=None):
+        examples.role_description.role_playing_with_role_description.main(
+            ModelType.GPT_3_5_TURBO)
```
Comment (lightaime) [inline examples/test/test_role_description.py:29]: Testing these two examples with `ModelType.GPT_3_5_TURBO` is too slow and expensive. Should we just patch the return and test them with `ModelType.STUB`?

File: camel/societies/role_playing.py [hunk 17]
```diff
@@ -233,12 +235,41 @@ def get_sys_message_info(
             initial system message, and a list of system message meta dicts.
         """
         sys_msg_meta_dicts = [dict(task=self.task_prompt) for _ in range(2)]
-        if (extend_sys_msg_meta_dicts is None and self.task_type
-                in [TaskType.AI_SOCIETY, TaskType.MISALIGNMENT]):
+        if (extend_sys_msg_meta_dicts is None and self.task_type in [
+                TaskType.AI_SOCIETY,
+                TaskType.MISALIGNMENT,
```
Comment (lightaime) [inline camel/societies/role_playing.py:240]:
```suggestion
                TaskType.MISALIGNMENT,
                TaskType.ROLE_DESCRIPTION,
```

File: examples/role_description/role_playing_with_role_description.py [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+    model_config_description = ChatGPTConfig()
+    role_description_agent = RoleAssignmentAgent(
+        model=model_type, model_config=model_config_description)
+
+    role_names, role_description_dict, _, _ = (
+        role_description_agent.run_role_with_description(
+            task_prompt=task_prompt, num_roles=2))
+
+    ai_assistant_role = role_names[AI_ASSISTANT_ROLE_INDEX]
+    ai_user_role = role_names[AI_USER_ROLE_INDEX]
```
Comment (lightaime) [inline examples/role_description/role_playing_with_role_description.py:38]: How do we know which is the assistant role and which is the user role?

File: camel/agents/role_assignment_agent.py
Comment (lightaime) [inline camel/agents/role_assignment_agent.py:1]: We may want to add a test for the agent.

File: camel/prompts/role_description_prompt_template.py
Comment (lightaime) [inline camel/prompts/role_description_prompt_template.py:1]: We may want to add a test for the prompt like https://github.com/camel-ai/camel/blob/master/test/prompts/test_ai_society.py.

**Review by lightaime — COMMENTED** [review]
Thanks, @Appointat. This is very cool. Left some comments. Let me know if there is any question!

**Comment by lightaime** [comment]
Thanks @Appointat @ZhangT-tech for adding this awesome future!!

**Pull request merged**